When byteorder record is missing load as little endian by default

AlekseiNikiforovIBM opened the pull request:
Fixes #101688

AlekseiNikiforovIBM committed:
When byteorder record is missing load as little endian by default

Fixes #101688

mikaylagawarecki reviewed (approved):
Following discussion here, I think it makes sense to approve this change and cherry pick it into the release branch as a critical fix to a new feature https://github.com/pytorch/pytorch/issues/101688#issuecomment-1701308630

mikaylagawarecki reviewed (commented):
Could you perhaps add a small test to make sure the behavior/warning is raised as expected

AlekseiNikiforovIBM commented:
Sure. Are there any tests for warnings I could use as reference?

AlekseiNikiforovIBM committed:
Add test for warning

Also change warning type from DeprecationWarning
to UserWarning to make it visible by default.

mikaylagawarecki commented on test/test_serialization.py:
```diff
@@ -3859,6 +3860,53 @@ def test_serialization_load_bom_data_cfloat(self):
+
+        buf_be_no_bom = io.BytesIO(data_be_no_bom)
+
+        with warnings.catch_warnings(record=True) as warns:
```
tiny nit: Here is an example of of how we usually catch warnings in test

https://github.com/pytorch/pytorch/blob/0f1a225f33b6e3966eda02baec5154fdc39a18dd/test/test_nn.py#L2703

mikaylagawarecki reviewed (commented):
Feel free to merge after you fix the nits and CI is green :)

After this is merged could you please open a PR to cherry pick this into the release branch. See instructions here
https://github.com/pytorch/pytorch/issues/108055. The criteria category should be "Critical fixes to new feature"

mikaylagawarecki commented on torch/serialization.py:
```diff
@@ -1326,27 +1326,27 @@ def _load(zip_file, map_location, pickle_module, pickle_file='data.pkl', overall
                       "torch.serialization.set_default_load_endianness to set "
                       "the desired default load endianness",
-                      DeprecationWarning)
+                      FutureWarning)
```
nit: `UserWarning`

mikaylagawarecki commented:
Looks like lint is failing

To fix the failing lint job running `make setup_lint` and `make quicklint` from your local pytorch directory and following the corresponding instructions should be sufficient, see [here](https://github.com/pytorch/pytorch/blob/4721553431e968ee5e90753346d92252cce98e84/Makefile#L25-L34)
